src/github.com/matrix-org/dendrite/roomserver/api/input.go

```diff
@@ -49,77 +46,66 @@ const DoNotSendToOtherServers = ""
 	// These are only used if HasState is true.
 	// The list can be empty, for example when storing the first event in a room.
-	StateEventIDs []string
+	StateEventIDs []string `json:"state_event_ids"`
 	// The server name to use to push this event to other servers.
 	// Or empty if this event shouldn't be pushed to other servers.
-	SendAsServer string
+	SendAsServer string `json:"send_as_server"`
 }
 
-// UnmarshalJSON implements json.Unmarshaller
-func (ire *InputRoomEvent) UnmarshalJSON(data []byte) error {
-	// Create a struct rather than unmarshalling directly into the InputRoomEvent
-	// so that we can use json.RawMessage.
-	// We use json.RawMessage so that the event JSON is sent as JSON rather than
-	// being base64 encoded which is the default for []byte.
-	var content struct {
-		Kind          int
-		Event         *json.RawMessage
-		AuthEventIDs  []string
-		StateEventIDs []string
-		HasState      bool
-		SendAsServer  string
-	}
-	if err := json.Unmarshal(data, &content); err != nil {
-		return err
-	}
-	ire.Kind = content.Kind
-	ire.AuthEventIDs = content.AuthEventIDs
-	ire.StateEventIDs = content.StateEventIDs
-	ire.HasState = content.HasState
-	ire.SendAsServer = content.SendAsServer
-	if content.Event != nil {
-		ire.Event = []byte(*content.Event)
-	}
-	return nil
+// InputRoomEventsRequest is a request to InputRoomEvents
+type InputRoomEventsRequest struct {
+	InputRoomEvents []InputRoomEvent `json:"input_room_events"`
+}
+
+// InputRoomEventsResponse is a response to InputRoomEvents
+type InputRoomEventsResponse struct{}
+
+// RoomserverInputAPI is used to write events to the room server.
+type RoomserverInputAPI interface {
+	InputRoomEvents(
+		request *InputRoomEventsRequest,
+		response *InputRoomEventsResponse,
+	) error
 }
 
-// MarshalJSON implements json.Marshaller
-func (ire InputRoomEvent) MarshalJSON() ([]byte, error) {
-	// Create a struct rather than marshalling directly from the InputRoomEvent
-	// so that we can use json.RawMessage.
-	// We use json.RawMessage so that the event JSON is sent as JSON rather than
-	// being base64 encoded which is the default for []byte.
-	event := json.RawMessage(ire.Event)
-	content := struct {
-		Kind          int
-		Event         *json.RawMessage
-		AuthEventIDs  []string
-		StateEventIDs []string
-		HasState      bool
-		SendAsServer  string
-	}{
-		Kind:          ire.Kind,
-		AuthEventIDs:  ire.AuthEventIDs,
-		StateEventIDs: ire.StateEventIDs,
-		Event:         &event,
-		HasState:      ire.HasState,
-		SendAsServer:  ire.SendAsServer,
+// RoomserverInputRoomEventsPath is the HTTP path for the InputRoomEvents API.
+const RoomserverInputRoomEventsPath = "/api/roomserver/inputRoomEvents"
+
+// NewRoomserverInputAPIHTTP creates a RoomserverInputAPI implemented by talking to a HTTP POST API.
+// If httpClient is nil then it uses the http.DefaultClient
+func NewRoomserverInputAPIHTTP(roomserverURL string, httpClient *http.Client) RoomserverInputAPI {
```

Comment:
It started becoming too much effort to get working and I'd like to focus on getting the matrix-specific logic written down. So for now it's much easier to stick with simple HTTP APIs.

It still should be fairly easy to switch out when we need to, I think the work is just:

1) Write protocol buffer schema for all the Query and Input RPC functions.
2) Set up a build environment with correct versions of protoc and grpc to do the codegen.
3) Setup prometheus style metrics for grpc.

But getting all that working and debugged could easily eat a couple of days, so I'd rather keep making progress on the core matrix stuff